snapcraft/tests/sources/test_git.py

```diff
@@ -383,3 +385,36 @@ def test_git_details_tag(self):
 
         self.source_details = self.git._get_source_details()
         self.assertEqual(self.expected_tag, self.source_details['tag'])
+
+
+class GitGenerateVersionTestCase(tests.TestCase):
+
+    scenarios = (
+        ('only_tag', dict(return_value='2.28',
+                          expected='2.28',
+                          tag=True)),
+        ('tag+commits', dict(return_value='2.28-28-gabcdef1',
+                             expected='2.28+git28.abcdef1',
+                             tag=True)),
+        ('tag+dirty', dict(return_value='2.28-29-gabcdef1-dirty',
+                           expected='2.28+git29.abcdef1-dirty',
+                           tag=True)),
+        ('no-tag', dict(return_value='abcdef1',
+                        expected='0+git.abcdef1',
+                        tag=False)),
+    )
+
+    def setUp(self):
+        super().setUp()
+
+        patcher = mock.patch('subprocess.check_output')
+        self.output_mock = patcher.start()
+        self.addCleanup(patcher.stop)
+
+    def test_version(self):
+        if self.tag:
+            self.output_mock.return_value = self.return_value.encode('utf-8')
+        else:
```

Comment:
I would prefer to move the no-tag case to a different test, to avoid the `if` block here.